pkg/frontend/querymiddleware/querysharding.go

```diff
@@ -270,6 +273,12 @@ func (s *querySharding) getShardsForQuery(ctx context.Context, tenantIDs []strin
 	maxShardedQueries := validation.SmallestPositiveIntPerTenant(tenantIDs, s.limit.QueryShardingMaxShardedQueries)
 	hints := r.GetHints()
 
+	if estimate := hints.GetEstimatedSeriesCount(); s.maxSeriesPerShard > 0 && estimate > 0 {
+		// If an estimate for query cardinality is available, use it to limit the number
+		// of shards based on linear interpolation.
+		totalShards = util_math.Min(totalShards, int(estimate/s.maxSeriesPerShard)+1)
```

Comment:
See line 304. Add a similar log here if the number of shards before/after has changed.